src/App.tsx

```diff
@@ -1,39 +1,65 @@
-import { useEffect, useState } from 'react';
 import React from 'react';
 import './App.css';
-import TodoList from './TodoList';
-import Todo, { Todos } from './types';
-
-export const TodosContext = React.createContext<Todos>([]);
+import Todo, { Todos, newTodoType } from './types';
+import { useQuery, useMutation } from 'react-query';
+import axios from 'axios';
+import { QueryClient } from '@tanstack/react-query';
 
+const fetchTodos = (): Promise<Todos> =>
+  axios.get('https://jsonplaceholder.typicode.com/todos').then((res) =>
+    res.data.reduce((acc: Todos, cur: Todo) => {
+      return [...acc, [cur.id, cur.title]];
+    }, [])
+  );
 
 function App() {
-  const [items, setItems] = useState<Todos>([]);
-
-  useEffect(() => {
-    fetch('https://jsonplaceholder.typicode.com/todos')
-      .then((response) => response.json())
-      .then((data) => {
-        const array = data.reduce((acc: Todos, cur: Todo) => {
-          return [...acc, [cur.id, cur.title]];
-        }, []);
-        setItems(array);
-      });
-  }, []);
-
-  const handleGenerateItem = () => {
-    setItems([...items, [items.length + 1, 'random text']]);
-  };
+  const queryClient = new QueryClient();
+
+  const { data, isLoading, error } = useQuery({
+    queryKey: ['todos'],
+    queryFn: fetchTodos,
+  });
+
+  const mutation = useMutation({
+    mutationFn: (newTodoItem: newTodoType) => {
+      return axios.post(
+        'https://jsonplaceholder.typicode.com/todos',
+        newTodoItem
+      );
+    },
+    onSuccess: () => {
+      queryClient.invalidateQueries({ queryKey: ['todos'] });
+    },
+  });
+
+  if (isLoading) return <div>Loading ...</div>;
+  if (error) return <div>Request failed with status code 404</div>;
+
+  const todosData = Array.isArray(data) ? data : [];
```

Comment:
You don't need this check. Just make sure the status is "success".

Reply:
Done